### ISSUE TITLE
Proposal for new core digest algorithm: BLAKE2 instead of MD5

### ISSUE THREAD

**Issue by hellyberry** [adrianlopezroche/fdupes#86]
Maybe the BLAKE2-algorithm solves both needs in software, fast and secure. It is 256 bit in size, cryptographically sound, so low change of 'wrong' collisions and relatively high speed.  Maybe for examining MD5 potential collisions, one can create a BLAKE2 digest of the files in question, this should be quicker than byte-by-byte comparison and is cryptographically secure.  See http://blake2.net

**Comment by hellyberry** [comment]
And the good thing, it exists in parallelized code for SMP systems exists, so running 4/8 threads is possible. This should speed up quite a bit.

> QUOTE: BLAKE2 is a cryptographic hash function faster than MD5, SHA-1, SHA-2, and SHA-3, yet is at least as secure as the latest standard SHA-3. BLAKE2 has been adopted by many projects due to its high speed, security, and simplicity.

BLAKE2 is specified in RFC 7693, and our code and test vectors are available on GitHub, licensed under CC0 (public domain-like). BLAKE2 is also described in the 2015 book The Hash Function BLAKE.

BLAKE2 comes in two flavors:

    BLAKE2b (or just BLAKE2) is optimized for 64-bit platforms—including NEON-enabled ARMs—and produces digests of any size between 1 and 64 bytes
    BLAKE2s is optimized for 8- to 32-bit platforms and produces digests of any size between 1 and 32 bytes 

BLAKE2 includes the 4-way parallel BLAKE2bp and 8-way parallel BLAKE2sp designed for increased performance on multicore or SIMD CPUs. BLAKE2 offers these algorithms tuned to your specific requirements, such as keyed hashing (that is, MAC or PRF), hashing with a salt, updatable or incremental tree-hashing, or any combination thereof. These versions are specified in the BLAKE2 document.

BLAKE2 also includes the BLAKE2x variants, which can produce digests of arbitrary length. BLAKE2x is specified in a separate document.

BLAKE2 shines on 64-bit CPUs: on an Intel Core i5-6600 (Skylake microarchitecture, 3310MHz), BLAKE2b can process 1 gibibyte per second, or a speed rate of 3.08 cycles per byte.

The plot below shows how BLAKE2 outperforms MD5, SHA-1, SHA-2, and SHA-3 on a Skylake Intel CPU (speeds are for hashing using a single core; using multiple cores, BLAKE2 can be even faster).
(from http://blake2.net/)

**Comment by jbruchon** [comment]
A secure hash algorithm is not needed since `fdupes` only uses the hash for exclusionary comparison. Collisions do not affect the final file match output. For this purpose, my [jodyhash algorithm](https://github.com/jbruchon/jodyhash) I wrote for the [jdupes](https://github.com/jbruchon/jdupes) fork is the fastest replacement option, even faster than xxhash, murmur3, etc.

**Comment by hellyberry** [comment]
I see. So fdupes and jdupes work same in this regard, but are different in the hashing-algorithm used.
What do you mean by exclusionary comparison?
Is first the file-size checked, then the MD5 (in fdupes)/jodyhash (in jdupes) if a file has same (non-zero) size?

**Comment by jbruchon** [comment]
`jdupes` is a fork of `fdupes`. Read the code for details. [jodyhash replacing MD5 was one of the first things I changed in the code.](https://github.com/jbruchon/jdupes/commit/f99ae1476bbc67733f8618ac6a301db159d2f18d) The benchmarked speed boost on a six-core machine was 17% over MD5 but Valgrind showed that the total CPU instructions executed dropped by 73% and in-memory data accesses dropped to one-third of the numbers with MD5. [Consolidating stat() call wrappers made a noticeable difference too](https://github.com/jbruchon/jdupes/commit/10c612e1a42eed1c4386dc4d4bc2d00e74a7344c) so switching hash algorithms isn't necessarily the only option for squeezing more speed out of things.

The basic comparison algorithm steps are:
- If sizes aren't equal, no match
- If permissions differ and the permission option was passed, no match
- If files are hard linked and `-H` not passed, no match, else yes match
- Hash first block of each file and compare. If different, no match
- Hash entire files and compare. If different, no match
- Compare files byte-by-byte. If different, no match, else yes match

Regarding an exclusionary comparison: if two pieces of data of any size are put through a one-way hash function and the resulting pair of hashes is different, the two pieces of data are guaranteed to be different as well. If the hash function produces the same hash in both cases, there is no guarantee that the data is identical OR different. Identical hashes can be the result of a collision. This is why `fdupes` (and `jdupes` by extension) has a final check which compares every byte of every file to guarantee that matches are really true 100% matches. This final check makes the randomness of the output of the chosen hash function less relevant because the worst case is that more files are unnecessarily read and compared until the first byte that is different.

Relying on the hash function alone to match files increases the risk of data loss, even with "secure" hash functions that seem to always produce a different result for different data. Refer to the [birthday problem](https://en.wikipedia.org/wiki/Birthday_problem) for a detailed explanation. As an example, SHA-512 has a 512-bit (64-byte) wide output. As soon as your data size fed to the hash algorithm exceeds 512 bits, you are guaranteed to have a chance of hash collision even if the hash function is extremely "secure." The real number is even lower because the data size is variable as well. For all practical purposes, this means that there is no hash algorithm which can guarantee no collisions for all possible inputs, which in turn means that collisions can and will happen, even if they are rare for most data sets fed to the program.

In other words, any one-way hash function will produce the same hash value if fed an identical input, so differing values mean difference in data, but identical values cannot be trusted due to the birthday problem dictating that hash collisions can and will occur, so we can use hashes to reliably detect data differences but NOT to detect data sameness.

By the way, if you are interested in my `jodyhash` algorithm, it was made with 64-bit CPUs in mind and the algorithm was explicitly designed to compile directly down to a small tight loop of native CPU instructions which is why it blows all other non-secure hash algorithms out of the water.

**Comment by kilobyte** [comment]
MD5 is so broken it allows computational complexity attacks: that is, a malicious user who can give you a bunch of files can force fdupes into comparing a large number of files pair-by-pair.  These files will be crafted to have the difference only on the last block.

On the other hand, while it is **theoretically** possible for a non-broken hash to produce collisions, comparing the files is totally pointless.  The definition of a "broken" hash is one where you can produce collisions cheaper than via brute force (ie, the birthday attack).  BLAKE2 (512 bit) has birthday length of 256 bits — that is, to have a fair chance to produce one collision, you'd need a computer with temp memory that stores 1 bit on every nucleon in the observable universe; in this case you can expect around 1 collision.  Half-length — that is — 256-bit version of BLAKE2 lowers the requirement to about 1 bit per nucleon in the mass of Earth.  I'm quite certain there's still a good few years before such storage capabilities will be plausible.

This is insanely less likely than the chance you'll be hit by three different lightning bolts out of clear sky today.

Any currently available computer also suffers from cosmic rays, thermal glitches and so on.  Any, no matter how correct, algorithm, will make mistakes because of CPU errors.

Thus, until BLAKE2 is broken, it's irrational to re-check the files.  Also, the recent SHA1 break required 5000 core-years to produce a single collision, so there's plenty of warning beforehand.

Yet another reason is that the current most important use case for fdupes, FIDEDUPERANGE, already does the byte-by-byte comparison inside the kernel (to avoid otherwise unavoidable races), thus repeating that work in userspace merely wastes time.

**Comment by FabioPedretti** [comment]
Blake3 is even better, faster and highly parallelizzable on multi core CPUs:
https://github.com/BLAKE3-team/BLAKE3
